### DISCUSSION TITLE
JSONB orderBy type

### DISCUSSION THREAD

**Discussion by JulianCissen** [mikro-orm/mikro-orm#4877] [Ideas]
My data model contains a JSON field (JSONB type in postgres database). Some of the elements in this JSON object can be strings formatted as dates/datetimes. This is not known at compile time, since the typing of the field is a simple `Record<string, unknown>`. The expected formatting is however known at runtime.

When I use the orderBy option using the EM find method to query my data, and specify the results to be ordered by one of these date/datetime fields, obviously it sorts them as if they are regular strings.

In postgres, and I assume other databases too, you can cast these fields in the order by clause, or apply a function to them there. This results in proper date sorting. Currently, the find methods of the EM don't support this type casting as far as I'm aware.

Since the types aren't known at compile time, I can't define a formula in my model that refers to this field, since I don't know the right key name yet. Also, a Formula decorator is a property decorator, not a method decorator, and I'm unsure if it would even be possible to implement this as some kind of method decorator where the method would supply the key to be used in the formula defined in the decorator. I'm not familiar enough with decorators for that.

There might be different approaches that could be done to solve this:
* A function expression could be allowed as orderBy value that returns a formula. I can imagine extending the concept of formula's beyond it's current scope could be unwanted though, since it's inherently not type-safe.
```typescript
em.find(Author, {}, {
  orderBy: () => `to_timestamp("author"."jsonbField"->>'birthdate', 'Dy Mon DD YYYY HH24:MI:SS') ASC`
});
```
* The value in the orderBy object could optionally be a combination of ASC/DESC and an optional property containing a type formatting hint. This specific implementation example would obviously limit the allowed formats since casting is more limited than the to_timestamp function.
```typescript
em.find(Author, {}, {
  orderBy: {
    jsonbField: {
      ['birthdate']: {
        queryOrder: 'ASC'
        type: DateType
      }
    }
  }
});
```
This could produce the query:
```sql
select * from author order by "jsonbField"->>'birthdate'::date ASC;
```

I'm curious:
* Has anyone had a similar use-case, and maybe solved it differently?
* Are there objections or inherent flaws to the design of any of the solutions above?

**Comment by B4nan** [comment]
You should be already able to do this (the `expr` helper is not really doing anything, it just lets the type checks pass):

```ts
em.find(Author, {}, {
  orderBy: { [expr(`to_timestamp("author"."jsonbField"->>'birthdate', 'Dy Mon DD YYYY HH24:MI:SS')`]:  'ASC' }
});
```

In v6 this will require `raw` helper (https://github.com/mikro-orm/mikro-orm/pull/4197).

**Reply by JulianCissen** [reply]
Ha, I did not know about this feature. Thank you for your help!

**Reply by B4nan** [reply]
FYI this is actually documented. 

https://mikro-orm.io/docs/entity-manager#using-custom-sql-fragments

And v6 adds a new section about raw queries, but that's not yet merged to master so not available in the docs:

https://github.com/mikro-orm/mikro-orm/pull/4197/files?w=1#diff-612b3c84e8ca654016789af2ad54d26a5c586fd64209c3c6730c2cc660129cb3

**Reply by JulianCissen** [reply]
I skimmed through this page a couple times, should have been more thorough :(

**Reply by JulianCissen** [reply]
Just to be sure:
I understand that `expr` currently just passes the string as SQL. I understand this makes it vulnerable to SQL injection when you pass user-input to the expr parameter.

In my use case, the user would supply the key in the json field on which ordering should take place. Resulting in something along the lines of this:
```typescript
em.find(Author, {}, {
  orderBy: { [expr(`to_timestamp("author"."jsonbField"->>'${keyFromFrontend}', 'Dy Mon DD YYYY HH24:MI:SS')`]:  'ASC' }
});
```

Meaning I'd have to do some real thorough validation to make sure it's not vulnerable.

The commit you mentioned which introduces the raw helper in v6 mentions it adds protection against SQL injection. But I think I understand correctly that it only protects against SQL injection from the parameter supplied as value to the SQL snippet, right? So the  `'ASC'` part in my example query (which there obviously is static, but in different use cases could be user defined), and not the `${keyFromFrontend}` part. Meaning I still have to extensively validate the input here.

**Reply by B4nan** [reply]
Yes, you need to validate things from users, especially those you want to include as part of a raw query. 

> The commit you mentioned which introduces the raw helper in v6 mentions it adds protection against SQL injection.

It protects you from passing user input as part of a regular find query, which could be interpreted as raw SQL. This is different, you say you basically want it there - and v6 will force you to explicitly allow it with the `raw` helper, otherwise the query (which works fine in v5) would fail, resulting in the whole key being wrapped in quotes. This is mostly about the keys, but applies to values as well, e.g. in v5 you can do `em.find(User, { [expr('lower(email)']: ... })`, and the `lower(email)` could come in from outside - v5 will allow it, while in v6 you need to be explicit and use the `raw` helper to make it work.

FYI the `raw` helper also allows you to use `??` in place of a key (for different quoting, as usually keys use `"` and values use `'`'). In v6 you could do something like the following:

```ts
em.find(Author, {}, {
  orderBy: {
    [raw(`to_timestamp(??->>?, ?)`, ['author.jsonbField', 'fooBar', 'Dy Mon DD YYYY HH24:MI:SS'])]:  'ASC',
  },
});
```

This will resolve to the same query, interpolating things as usual, so something like `to_timestamp("author"."jsonbField"->>'fooBar', 'Dy Mon DD YYYY HH24:MI:SS')`.

**Reply by JulianCissen** [reply]
I see, I think the support for the unnamed parameters in the `raw` helper in v6 is crucial for me to implement appropriate SQL injection protection. Thank you for your help!

**Reply by B4nan** [reply]
Technically, you can go for v6 right now, every commit is published to NPM already, and right now I would call it almost ready - I am leaving for a few days now, if it wasnt's like that, I would probably merge it to master during weekend. I think the stable will go out in the end of the year, but the main focus now will be on the docs and bug fixes, I am pretty much done with all the code changes (there is one last feature I want to implement, and don't plan any more breaking changes or bigger internal refactors).

**Reply by JulianCissen** [reply]
Thanks for the suggestion, but however mundane my examples, my actual use is in a commercial product so I can't use non-"stable" versions. There's no rush for this functionality right now so I can wait patiently.
